### PR TITLE
Isolating core parameterization and architecting for recoupling

### LINKED ISSUE #9: Prune / refactor / modularize NN_module code
The code in `NN_module/` has currently been ripped out of it's original implementation in SAM along with any dependencies.
However, this likely contains much redundant code, so we should investigate and try to pare it down a bit.

### LINKED ISSUE #10: NN_module build process
Currently there is a shell script for building the module in `NN_module/` on CSD3.
However, this should really be reformulated as a Makefile.
Could also consider CMake.

### LINKED ISSUE #35: Add build, run and test documentation to readme
Add build, run and test documentation to readme so that the process of building and running the code is clear.

Provide some justification to lack of CMake if we decide not to use it.

### LINKED ISSUE #36: Establish dead code fragments
Whilst refactoring the code as part of #33 there are a few questions that have come up.

Looking at [Janni's original repo](https://github.com/yaniyuval/Neural_nework_parameterization/tree/v.1.0.3/sam_code_NN) there are a number of subdirectories under [`sam_cases/`](https://github.com/yaniyuval/Neural_nework_parameterization/tree/v.1.0.3/sam_code_NN/sam_cases), each with a different version of `nn_convection_flux.f90`, `nn_diffusion.f90`, and parameters file `prm`.

It seems that the code was designed to test a variety of configurations with/without certain variables during development through the use of switches such as `rf_uses_rh` etc. Is this behavior that needs preserving, or is there a 'final'/'best' version of the net that we are wanting to implement?

If we only want one version of the net then we can(should) remove dead code fragments to speed things up and simplify the code.

If we are wanting to implement a specific net, then which one of the instances in [`sam_cases/`](https://github.com/yaniyuval/Neural_nework_parameterization/tree/v.1.0.3/sam_code_NN/sam_cases) is it, or is it the [`nn_convection_flux.f90`](https://github.com/yaniyuval/Neural_nework_parameterization/blob/v.1.0.3/sam_code_NN/nn_convection_flux.f90) file in the `sam_code_NN/` directory?

If the latter then is this governed by the parameters listed as defaults in [`setparm.f90`](https://github.com/yaniyuval/Neural_nework_parameterization/blob/v.1.0.3/sam_code_NN/setparm.f90), as there is no `prm` parameters file in the main directory?

Finally I note that there is a file [`nn_diffusion.f90`](https://github.com/yaniyuval/Neural_nework_parameterization/blob/v.1.0.3/sam_code_NN/nn_diffusion.f90). Do we need to do anything with this? My understanding was that we were extracting `nn_convection_flux.f90`, but perhaps I have misunderstood?

### LINKED ISSUE #40: Rename t_i to tabs_i
As discussed with @paogorman SAM uses `tabs` for absolute temperature, and `t` for moist static energy.

However, the parameterisation features a variable `t_i` which represents absolute temperature at another point in the time cycle (`_i` used similarly for other variables).

We should refactor so that `t_i` becomes `tabs_i` to be consistent and scientifically clearer.

### PR DESCRIPTION
**Pull request by jatkinson1000**
This PR should address and close #9 and close #10 and close #36 and close #40 and maybe close #35
- [X] Remove dependence on various SAM modules
- [X] Separate NN into separate functions (adressed in #23 and #28 which are merged here)
- [x] Clarify what is required as an input to the NN, and what needs to come from the model
- [x] Create a model specific wrapper function to feed into the NN - this can then be adapted for CAM/SCAM
- [x] Clarify physics processing of outputs from NN in `nn_convection_flux`

It also takes steps towards addressing #10 by adding a Makefile.
However, there should be discussion over the build process we wish to use.
The makefile currently requires a user to specify the location of NetCDF libraries to link against.

### CODE REVIEW

**Comment by dorchard** [comment]
There are some big commented out blocks, I think due to the us dead-code removing code paths that won't be used because`rf_uses_rh` is `.false.`. Should we remove these in this PR totally then?

**Comment by jatkinson1000** [comment]
> There are some big commented out blocks, I think due to the us dead-code removing code paths that won't be used because`rf_uses_rh` is `.false.`. Should we remove these in this PR totally then?

I think the block can perhaps be removed from the SAM interface code, but am inclined to leave them in the parameterisation as it's one main block and functions at the end.
Perhaps something to ask Paul on Thursday about how serious they are to re-implement in future?
Other option would be to open a separate issue to remove `rf_uses_rh` code so that we can easily find the commit where it was present in future.

**Comment by jatkinson1000** [comment]
I have implemented the 3-layer structure, cleaned the code, and got rid of any implicit dependency on SAM.

These is still some physics to clarify, but this does not affect operation, so I'd like review on this now to merge and move to interfacing with SAM to check it still works OK.
This will also allow @raehik to look into CamFORT etc.

Moving from draft to review.

**Comment by jatkinson1000** [comment]
This PR will also close #10

**Review by raehik — APPROVED** [review]
LGTM. Reads nicely. Some changes to the Makefile to get things compiling on most easy platforms.

File: NN_module/nn_cf_net.f90 [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        out_dim_counter = out_dim_counter + nrf-1
+
+        i = i + 1
+        logits(out_dim_counter+1:out_dim_counter+nrf) = (logits(out_dim_counter+1:out_dim_counter+nrf) * yscale_stnd(i))  +  yscale_mean(i)
+        out_dim_counter = out_dim_counter + nrf
+
+        i = i + 1
+        logits(out_dim_counter+1:out_dim_counter+nrf) = (logits(out_dim_counter+1:out_dim_counter+nrf) * yscale_stnd(i))  +  yscale_mean(i)
+        out_dim_counter = out_dim_counter + nrf
+
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_cf_net.f90:139]: Noice, will add.

[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_cf_net.f90]: Took a bit of rejigging of the above (including naughty use of `size` as a variable!) but in now and seems to still produce same numbers.

File: NN_module/nn_convection_flux.f90 [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+
+        ! Set init flag as complete
+        do_init = .false.
+
+        ! These have been made module parameters
+        ! Hardcoded magic numbers
+        ! nrf is the fact results are supplied at lowest 30 half-model levels for sedimentation fluxes, and at 29 levels for fluxes
+        ! (as flux at bottom boundary is zero).
+        ! nrf = 30
+        ! nrfq = 29
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_convection_flux.f90:114]: Done

File: NN_module/nn_convection_flux.f90 [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+                t(i,j,1:nrf) = t(i,j,1:nrf) + t_rad_rest_tend(1:nrf)*dtn
+
+                ! Calculate surface precipitation
+                ! Apply sed flux at surface
+                precsfc(i,j) = precsfc(i,j)  - q_flux_sed(1)*dtn*rev_dz ! For statistics
+                prec_xy(i,j) = prec_xy(i,j)  - q_flux_sed(1)*dtn*rev_dz ! For 2D output
+                !
+                do k=1, nrf
+                    ! TODO: Both these lines have dz*(1/dz) can we just remove this?
+                    precsfc(i,j) = precsfc(i,j) - q_tend_tot(k)*adz(k)*dz*rho(k)*rev_dz! removed the time step mult because q_tend_tot is already mult
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_convection_flux.f90:388]: Done.

File: NN_module/nn_convection_flux.f90 [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+        ! -----------------------------------
+        ! NN variables
+        real(4), dimension(n_inputs) :: features
+            !! Vector of input features for the NN
+        real(4), dimension(n_outputs) :: outputs
+            !! vector of output features from the NN
+
+!         real :: omn
+!             !! variable to store result of omegan function
+        real :: rev_dz
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_convection_flux.f90]: Don't think we can do it much earlier as `dz` is a function parameter so assignment needs to come after all other variable definitions? Leaving as is for now.

File: NN_module/nn_convection_flux.f90 [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+                                  t, q, qn, precsfc, prec_xy)
+        !! Interface to the neural net that applies physical constraints and reshaping
+        !! of variables.
+
+        ! -----------------------------------
+        ! Input Variables
+        real, intent(in) :: y_in(:)
+            !! Distance of column from equator (proxy for insolation and sfc albedo)
+        ! Fields from beginning of time step used as NN inputs
+        real, intent(in) :: t_i(:, :, :)
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_convection_flux.f90:124]: Yep, wanted to be very careful with this so just been through and done it for as much as I can. Have opened #40 to reflect this.

**Comment by jatkinson1000** [comment]
This will now close #40

**Comment by jatkinson1000** [comment]
Useful discussion with @paogorman who answered a number of questions I had around the science/numerics.
These will inform the refactor and as a result I will be making a few more (notable) changes on this PR so will likely want a re-review soon.
Summarised below:
* Questions:
  * `q` is total water, and `qn` is cloud water.  
    However, both are updated with adv + tot + sed, so what is the difference?  
    There are different checks on physical correctness (q checked at each addition, qn just checked after all added)?
    - `q` is total water content (vapour plus `qn`), `qn` is non-vapour water content.
    - We can probably comment out `qn` as it comes from SAM setup - in larger models it will be immediately precipitated out during a larger timestep.
    - In fact, just delete `qn`!!
  * gt lt misses 0.0 -> can we add?
    - Think about this, likely negligible effect from changing in reality
  * Check variable names and units
    - Seems OK
  * tendency is $\Delta\psi$?
    - Yes. Itr would be nicer if we can refactor for tendencies to actually be tendencies rather than changes
  * Flux is what related to `irhoadzdz`?
    - SAM equations are of the form:  
    $\text{tendency} = \text{d} h / \text{d}t = - 1 / \rho * \text{d} (\rho * u * h) / \text{d}x = - 1 / \rho * \text{d}(\text{flux}) / \text{d}x$
      So the `irhoadzdz` term comes from rearranging to get:
      $\text{d}h = - \text{irhoadzdz} * \text{d}({flux})$
  * What is `_tot`?
    - tot is a legacy from when the variable was total change. Now the tendencies are checked separately and this should really be called auto
    - CAM operates at the level of tendencies.
    - careful about removing dt as there is a dtn in irhodzdz

  - t is `energy / cp` beware of this, especially in the SAM-CAM conversion

File: NN_module/nn_convection_flux.f90 [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+                    else
+                        q_tend_tot(k) = q_tendency_auto(k) * dtn
+                    endif
+                end do
+
+                ! Update q and t
+                q(i,j,1:nrf) = q(i,j,1:nrf) + q_tend_tot(1:nrf)
+                t(i,j,1:nrf) = t(i,j,1:nrf) - q_tend_tot(1:nrf)*fac(1:nrf)
+
+                ! TODO ???
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_convection_flux.f90]: Sorted this - see comments in latest code.

**Comment by jatkinson1000** [comment]
This is a fairly sizeable PR with several linked issues at this point.
I propose we merge this back (after SAM integration check) and then open a subsequent PR linked to issues listed in the above comment refactoring further towards CAM interfacing by changing outputs to be tendencies rather than updated variables.
Thoughts @raehik @dorchard ?

**Pull request merged**